File: Explore/2020April/Week5/Python/binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        return self.sumRoot(root)[0]
    
    def sumRoot(self, node):
        if not node:
            return float('-inf'), 0
        
        leftVia, leftDown = self.sumRoot(node.left)
        rightVia, rightDown = self.sumRoot(node.right)
        via = max(node.val + max(0, leftDown) + max(0, rightDown), leftVia, rightVia)
        down = node.val + max(0, leftDown, rightDown)
        
        return via, down

File: Explore/2020April/Week5/C/binary-tree-maximum-path-sum.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int max3(int a, int b, int c)
{
    int max = a;
    
    if(max < b)
        max = b;
    
    if(max < c)
        max = c;
    
    return max;
}

int sumRoot(struct TreeNode* root)
{
    if(root == NULL)
        return 0;
    
    return max3(sumRoot(root->left) + root->val, 
                sumRoot(root->right) + root->val, 
                0);
}

int maxPathSum(struct TreeNode* root){
    if(root == NULL)
        return INT_MIN;
    
    return max3(
        sumRoot(root->left) + root->val + sumRoot(root->right), 
        maxPathSum(root->left), 
        maxPathSum(root->right));
}
